Rewrote the Slice Grammar Reference Page (#354)
icerpc/icerpc-docs: pages/slice/language-reference/slice-grammar.md

title: Slice grammar
---

> This page describes the grammatical rules, usage, and restrictions of the Slice language.

Author: Yes, I mentioned this in my original comment:
> I'm aware some lines overflow the 120 column limit. I'll format the file once everyone is okay with the content.
Letting them overflow for now makes it easier to review and offer code suggestions.

## Slice elements

### Slice files

The `SliceFile` grammar rule is the parser's entry point.
Once a file's contents have been [preprocessed], the resulting text is matched against this rule, in its entirety.

> While the contents of a Slice file may reference things defined in other files, each file is parsed separately.

Author: Changed to "may reference definitions in other files".

> There is no notion of 'including' the contents of one file in another, and errors in one file do not affect the parsing of others.

Author: Changed as suggested.

> A Slice file consists of an optional [module declaration][module-declaration], followed by any number of Slice definitions.

Author: An empty file is a valid Slice file.
In the case where you don't have any Slice definitions, you're allowed to omit the module too.
We actually mention this in the `module declaration` section:
> If a Slice file contains no Slice definitions, it is legal to omit the module declaration.
Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,

> Additionally, [file attributes][attribute] and a [mode declaration][mode-declaration] may be specified at the top of a file, before any other definitions.

Author: I've rewritten this section to be more clear:
```
A slice file consists of any number of [file attributes][attribute] and up to one [mode declaration][mode-declaration].
These can appear in any order, but declaring more than one mode in a single file is forbidden.
This is followed by an optional [module declaration][module-declaration], and then any number of Slice definitions.
```

Declaring more than one mode in a single file is forbidden.

```ebnf {% showTitle=false %}
SliceFile

>     : SliceFilePrelude Module? Definition*

Author: Yes, but Slice definitions are also optional.
An empty file is a valid Slice file.

    ;

SliceFilePrelude
    : EMPTY
    | SliceFilePrelude FileCompilationMode
    | SliceFilePrelude FileAttribute
    ;

Definition
    : Struct | Class | Exception | Interface | Enum | CustomType | TypeAlias
    ;
```

For additional information on Slice files, see the [slice file][slice-file-guide] page.

### Mode declarations

> Mode declarations consist of the `mode` keyword, followed by an equals sign, followed by an identifier.

Author: I've renamed this to the "mode statement" (both in my PR and in the compiler).

> Mode declarations consist of the `mode` keyword, followed by an equals sign, followed by an identifier.

Reviewer: I am really not fond of this wordy description. Just read the grammar below.

> Mode declarations consist of the `mode` keyword, followed by an equals sign, followed by an identifier.

Author: I think these comments were meant to go on https://github.com/icerpc/icerpc-docs/pull/354#discussion_r1332040051

This was about renaming "mode declarations" to "mode statements",
which I think is fine and have already changed.

That identifier must be a valid Slice compilation mode - either `Slice1` or `Slice2`.

```ebnf {% showTitle=false %}
FileCompilationMode
    : "mode" "=" Identifier
    ;
```

> Mode declarations must come before any Slice definitions, including [module declarations][module-declaration].

Author: Changed to "any Slice definitions or [module declarations]".

> While the compilation mode you select will affect which Slice features are available to you,

Author: Yes, I prefer your simpler structure.
But, saying that you're 'using' features sounds too active for a reference. Would you be okay with:
"The compilation mode you select determines which Slice features are available to you but has no effect on"?

> While the compilation mode you select will affect which Slice features are available to you,

Reviewer: Yes that's fine.

> While the compilation mode you select will affect which Slice features are available to you,

Author: Okay, fixed!

it has no effect on the parsing of, or validity of syntax within, your Slice file.
For example, even though classes are a Slice1-only feature, `class` is still a keyword within a Slice2 file,

> and if you define a class, syntax errors in that class's definition will still be reported normally.

Author: I think emphasizing that we're talking about that class is important here.

Otherwise the phrase could mean: "we report a syntax error for using a class",
when what it's conveying is: "we still fully parse the class for syntax errors".

For additional information on modes, see the [compilation mode][compilation-mode-guide] page.

### Module declarations

Module declarations consist of the `module` keyword, followed by an [identifier].

> [local attributes][attribute] can be applied to a module in its prelude, but [doc comments][doc-comment] on modules are forbidden.

Author: I think it does. The sentence shows a contrasting relationship between attributes and comments.
The element of contrast is "can this go on a module?".

For example:
"Alice can go to the mall, but John cannot" (maybe he's grounded)
"attributes can go on modules, but doc comments cannot"

> [local attributes][attribute] can be applied to a module in its prelude, but [doc comments][doc-comment] on modules are forbidden.

Reviewer: I disagree. Even without the passive voice, it's odd.

> [local attributes][attribute] can be applied to a module in its prelude, but [doc comments][doc-comment] on modules are forbidden.

Author: I don't see what's odd about it, and neither does anyone I've asked (only 2 people to be fair).
This is a correct and normal sounding English sentence as far as I can tell.

```ebnf {% showTitle=false %}
Module
    : Prelude "module" RelativeIdentifier
    ;
```

If the module's identifier is [unscoped][identifier], this declares a _top-level module_.
If the module's identifier is [scoped][identifier], this declares a _submodule_.

Submodules are modules that are logically contained within another module (called the _parent_).
Slice definitions within a submodule can reference types in parent modules without qualification.

```slice
module A // Declares a top-level module named `A`.
struct AStruct {}
```

```slice
module A::B // Declares a submodule named `B` that is contained within module `A`.
typealias T = AStruct // Don't need to qualify as `A::AStruct`.
```

> All slice definitions must be contained within modules - Slice definitions cannot exist at 'global' scope.

Author: Good catch. I fixed the capitalization of Slice. There were a couple other places I missed too.

Also removed the quotes.

If a Slice file contains no Slice definitions, it is legal to omit the module declaration.
Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,
and any Slice definitions in the remainder of the file will be contained within the declared module.

Module declarations must come before any Slice definitions and must come after any [file attributes][attribute] or [mode declarations][mode-declaration].
There can be at most one module declaration per Slice file.

```ebnf {% showTitle=false %}
SliceFile
    : SliceFilePrelude Module? Definition*
    ;
```

It is legal for multiple Slice files to declare the same module.
Slice definitions defined in those files will be in the same module, even if they're in different Slice files.

For additional information on modules, see the [module][module-guide] page.

### Primitive types

> Slice supports 17 primitive types (although `AnyClass` is only valid in [`Slice1`][compilation-mode-guide] mode).

Author: You're right. I removed this statement, and refactored the sentence to sound okay without it.

Primitive types are always available: they're defined in any scope of any Slice file.
They can be referenced with their keywords:

```ebnf {% showTitle=false %}
Primitive
    : "bool"      | "int8"  | "uint8"  | "int16"    | "uint16"    | "int32"   | "uint32"  | "varint32"
    | "varuint32" | "int64" | "uint64" | "varint62" | "varuint62" | "float32" | "float64" | "string"
    | "AnyClass"
    ;
```

For additional information on primitive types, see the [primitive type][primitive-type-guide] page.

### Collection types

> Slice supports 2 built-in collection types: sequences and dictionaries.

Author: Is this the terminology we're using elsewhere?
I really don't think we should be talking about _instantiating_ things in Slice.

Anyways, I think most languages consider generic _types_ to still be _types_.

You're correct that to be usable (right now), you have to supply concrete arguments to them (thereby converting a generic type into a constructed type) but they're still _types_.

Consider when we add generic structs.
```
struct Hello<T> {
    data: T,
}
```
We're going to document things like `Hello` under the `struct types` section.
And it's valid to ask what the _type_ of the field 'data' is, even though it's a generic field.

> Slice supports 2 built-in collection types: sequences and dictionaries.

Reviewer: We currently describe Sequence and Dictionary in a single spot, where we call them built-in generic types. I think we should use this terminology consistently.

I think "type" is a poor name. A Sequence is not a valid type. Your field's type can't be Sequence.

> Slice supports 2 built-in collection types: sequences and dictionaries.

Author: I understand the point you're making.
Yes, a 'generic type' isn't a _usable_ type. And to make them functional you need a 'constructed type'.
In this regard, 'type' is indeed a poor name.

But, it's the name we're stuck with. Every reader and other reference out there still calls these _types_.
I mean, `generic type` literally has the word `type` in it.
So it's just somewhat undeniable that these things are a _kind_ of type, even if not a very usable kind.

> Slice supports 2 built-in collection types: sequences and dictionaries.

Reviewer: I am just requesting that you use the same terminology as:
https://docs.icerpc.dev/slice2/language-guide/sequence-types

> Slice supports 2 built-in collection types: sequences and dictionaries.

Author: Even though we disagree on which is _more_ important, I think we can agree that sequences and dictionaries are both generic types, and collection types.

So, what if we say: `Slice has built-in support for 2 generic collection types: sequences and dictionaries`.

Both types can be referenced with their respective keywords: `Sequence` and `Dictionary`,

```ebnf {% showTitle=false %}
Sequence

>     : "sequence" "<" TypeRef ">"

Author: Good catch. Fixed.

    ;

Dictionary

>     : "dictionary" "<" TypeRef "," TypeRef ">"

Author: Also fixed.

    ;
```

> Since they are are generic over the types of elements they contain, you must also specify their type arguments:

Author: Rewritten to: `to be usable, they must have concrete types specified for their type parameters`

```slice
Sequence<bool>
Dictionary<uint8, string>
```

> Since dictionary keys must be comparable, there are restrictions on which types can be used for them.

Author: I didn't document it because we already talk about it in the language guide.

I've now added the restrictions to the reference, but should I leave the guide-version intact, or replace it with a link to this section of the reference?

> Since dictionary keys must be comparable, there are restrictions on which types can be used for them.

Reviewer: The Guide and Reference can occasionally duplicate information.

> Since dictionary keys must be comparable, there are restrictions on which types can be used for them.

Author: Alright, then I'll leave the guide as-is.
Please look through my rewrite of the sequence/dictionary section then.

For additional information on sequences and dictionaries, see the [sequence][sequence-guide] and [dictionary][dictionary-guide] pages.

### Struct types

> Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.

Author: The main advantage (for me) is that we can **link to things** in the textual version, which makes it easier to navigate the reference. Say you're reading about structs and want to know about fields.
It would be awesome having the EBNF tokens (in the code block) be clickable links, but this is hard to implement.

Also the textual version **does provide some additional information**.
The grammar rule `UndelimitedList` doesn't communicate that fields may be optionally separated by commas.
`Prelude` doesn't communicate that this can be attributes or a doc comment.
etc. You _can_ get this information by navigating to those rules, but mentioning them here makes it easier to read.
A reader doesn't need to hunt around for this extra information.

This is also a **standard practice**. It's formatted differently, but you can see the same idea in the C# reference:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1521-general

> Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.

Reviewer: I very much disagree. There is no point in just reading the EBNF grammar for you. If there is relevant additional information, sure, give this _additional_ information.

For example:
> Struct definitions consist of the struct keyword, followed by an [identifier], and then the struct's body. Struct bodies consist of a list of [fields][field] wrapped in a pair of braces. 

I don't see any information above that goes beyond a plain reading of the EBNF grammar.

As for the links, I would simply add:
See also: link1, link2, link3.

> Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.

Author: > I very much disagree.

What are you disagreeing with here?
Having these text blocks at all?
Or one of the reasons given by @externl or I?

----

> In any of the above NOT expressed by the ebnf grammar?
> I don't see any information above that goes beyond a plain reading of the EBNF grammar.

I already answered this in my prior comment. There _is_ information that goes beyond a plain reading:
> The grammar rule UndelimitedList doesn't communicate that fields may be optionally separated by commas.
Prelude doesn't communicate that this can be attributes or a doc comment.

The grammar rule does not convey this information, but the sentences do. For reference:
```
Struct
    : Prelude "compact"? "struct" Identifier "{" UndelimitedList<Field> "}"
    ;
```

Putting this information in context, with written sentences makes it easier to read,
instead of just linking to other parts of the specification.

I would also think a textual version of rules would be more digestible to some readers then
"Here's a grammar rule in an unfamiliar syntax"

### Type aliases

Type alias definitions consist of the `typealias` keyword, followed by an [identifier], followed by an equals sign, and then a [type].
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the type alias in its prelude.

```ebnf {% showTitle=false %}
TypeAlias
    : Prelude "typealias" Identifier "=" TypeRef
    ;
```

Type aliases can be used anywhere a type can be used, and behave identically to their underlying type.
It is legal for type aliases to alias other type aliases.

For additional information on type aliases, see the [type alias][type-alias-guide] page.

### Type references

Type references consist of any number of [type attributes][attribute], followed a type, optionally followed by a question mark.

Built-in types like [primitives][primitive], [sequences][sequence], and [dictionaries][dictionary] can be referenced by their keywords.
User defined types can be referenced by their (possibly scoped) [identifiers][identifier].

```ebnf {% showTitle=false %}
TypeRef
    : LocalAttribute* TypeRefDefinition "?"?
    ;

TypeRefDefinition
    : Primitive | Sequence | Dictionary | RelativeIdentifier | GlobalIdentifier
    ;
```

> ### Tags

Author: I'll address the structure of this once the PR is merged.

Tags consist of the `tag` keyword, followed by a positive integer wrapped in a pair of parenthesis.
This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).
Tags can only be applied to [fields][field] and [parameters][parameter], and only if the type of that field/parameter:
- is optional (it ends with `?` symbol)
- is not a class, and does not use classes internally

```ebnf {% showTitle=false %}
Tag
    : "tag" "(" SignedInteger ")"
    ;
```

### Identifiers

Slice supports three forms of identifiers:
- An unscoped identifier. These consist of a letter, followed by any number of letters, digits, or underscores.
- A relatively scoped identifier. These consist of one or more unscoped identifiers, separated by `::` tokens.
- A globally scoped identifier. These consist of a `::` token, followed by a relatively scoped identifier.

```ebnf {% showTitle=false %}
Identifier
    : identifier
    ;

RelativeIdentifier
    : identifier ("::" identifier)*
    ;

GlobalIdentifier
    : ("::" identifier)+
    ;
```

Examples of valid identifiers:

```slice
// Examples of unscoped identifiers
foo9

> _bar_

Author: Very good catch. Fixed.

BAZ

// Examples of relatively scoped identifiers
foo9

> foo9::_bar_::BAZ

Author: Also fixed. Not sure how I missed such a big mistake.

To use a keyword as an identifier, you can escape the keyword by prefixing it with a backslash.
For scoped identifiers, each individual identifier must be escaped this way.
It is legal to escape an identifier that is not a keyword, though doing so has no effect.

```slice
struct   // The struct keyword
\struct  // An unscoped identifier
\foo     // An unscoped identifier

::\struct::\module // A globally scoped identifier
```

### Integer literals

Slice supports three forms of integer literals:
- A decimal literal for base-10 numbers. These consist of one or more digits between `0` and `9`.
- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.

Additionally, underscores can appear between any two characters of an integer literal.
These underscores carry no semantic meaning and are discarded by the parser.

> Examples of valid integer literals:

Author: I think the opposite : v) The focus of this section is solely integer literals. Suddenly bringing `tag`s into it would just blur what's being defined here I worry.

Plus, it's still the same problem:
You can't just put an integer literal in the middle of your Slice file.
But you also can't just put a tag in the middle of your Slice file.

```slice
123
0xFF
0b1010

335_445_996
0x_ab_cd_ef  // Same as 0xabcded
0b0_______1  // Same as 0b01

0x_0b1101  // hexadecimal literal of value 725249

// All have a value of 0
0
0_00_0
0b000
0x0
```

{% callout type="note" %}
Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
{% /callout %}

The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.

```ebnf {% showTitle=false %}
Integer
    : integer_literal
    ;

SignedInteger
    : Integer
    | "-" Integer
    ;
```

> ### String literals

Reviewer: It would be useful to mention that string literals are used only for attribute parameters.

Author: Agreed, I added this to the section: `String literals can only appear as arguments to an [attribute].`

String literals consist of any number of UTF8 characters, wrapped in a pair of double quotes.
Any characters within a string literal are treated as opaque text by the parser.

> Slice supports escaping special characters within string literals, this allows you to include unprintable characters or literal double quote characters in your string. Characters can be escaped by prefixing them with a backslash.

Author: I split the sentence up, let me know if that fixed the mistake you saw.

To write a literal backslash, you must escape it with another backslash: `\\`.

> When parsing, Slice will resolve any escaped characters, removing the `\` escape character where present.

Reviewer: No need for "will". I assume you meant "the Slice parser" or "the parser".

Author: Yes. I simplified it to:
> The parser will resolve any escaped characters, removing the `\` escape character where present.

Apart from this, Slice performs no additional processing of string literals.

Examples of valid string literals:

> ```slice

Reviewer: Like for integer literals, it would be better to show actual use-cases - namely, attributes.

Author: I'm still worried this muddles the line of what this section is focused on.

If you're trying to write a Slice language server for Eclipse for instance, and you're writing the code for handling string literals, whether it's in an attribute or not doesn't matter. The Slice highlighters works this way too. They have a rule for highlighting strings that's totally indifferent to the context.

""
"hello world"
"the variable \"foo\" doesn't exist"
"a backslash: '\\'"
```

### Attributes

Slice supports two forms of attributes:

> - Local attributes: these consist of an attribute wrapped in a pair of single brackets.

Reviewer: My preference is to use the same terminology as on https://docs.icerpc.dev/slice2/language-guide/attributes, namely attributes and file-level attributes. No local attribute / file attribute.

Author: I didn't know we were just calling them 'attributes' now.

I've always found 'local' and 'global' attributes to be bad terms, but,
what would be the umbrella term for "attributes" and "file-level attributes" though?...
Writing "Slice supports two forms of attributes: attributes and file-level attributes" would be odd.

Reviewer: > what would be the umbrella term for "attributes" and "file-level attributes" though

Easy: we don't need an umbrella term. We have attributes and file attributes.

Author: Okay, just to make sure we agree on the concepts here, you're saying:

we only have attributes, and those attributes can on things.
`oneway` is an operation attribute, `cs::namespace` is a module attribute, and `cpp::header-ext` is a file attribute.

Slice just happens to require a slightly different syntax for file attributes? But it's still an _attribute_.

Author: I rewrote the whole attribute section.
We don't talk about local attributes anymore, we just talk about attributes.

And to apply an attribute to a file, use `[[`.
To apply an attribute to a definition/declaration, use `[`.
But they're all just _attributes_.

- File attributes: these consist of an attribute wrapped in a pair of double brackets.

> Local attributes can be applied to most Slice definitions by placing them directly before the definition.

Reviewer: I find this confusing: you can apply (local) attributes to module declarations.

> File attributes apply to an entire [slice file][slice-file] and must appear before any Slice definitions or module declarations.

Reviewer: You should always spell Slice with a capital S.

Author: Fixed.

```ebnf {% showTitle=false %}
FileAttribute
    : "[[" Attribute "]]"
    ;

LocalAttribute
    : "[" Attribute "]"
    ;
```

An attribute consists of a directive, optionally followed by attribute arguments.

> A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of Slice scopes, its scopes determine when the attribute is meaningful.

Reviewer: I find the comparison with Slice scopes misleading since we use different naming conventions: module names are by convention in Pascal case while attribute scopes are lower case.

Author: The conventions are different, but I don't think conventions are technically relevant in the reference.
Even if we suggest different conventions, they are still syntactically equivalent.

This sentence isn't about advertising "use these like identifiers",
it's for saying "we're not going to restate what this is, since it's the same as this other thing; click here"

> For example, a directive with no scope like `deprecated` will always apply, but a directive like `cs::type` will only apply when running `slicec-cs`.

Reviewer: This doesn't sound like the correct way to describe attribute "scopes". I think it's more like:
 - no scope => handled by the base parser (slicec)
 - some other scope => not handled by the base parser

 Also, keep in mind we are describing / should be describing the grammar here, not implementation details of the slicec library.

Attribute arguments consist of any number of arguments wrapped in a pair of parenthesis, separated by commas.
Arguments may be either a [string literal][string] or an [unscoped identifier][identifier].

```ebnf {% showTitle=false %}
Attribute
    : RelativeIdentifier ("(" CommaList<AttributeArgument> ")")?
    ;

AttributeArgument
    : string_literal | identifier
    ;
```

How many arguments are accepted/required is up to the directive.

> But supplying too many or too few arguments is never a syntax error.

Reviewer: I am confused by this "note". It's obviously an error to give 10 parameters to the deprecated attribute.
If we explain "too many / too few parameters is not the parser problem", we should also "note" whose problem it is.

Author: I just removed the "it's not a syntax error" part. It's now:
```
Each attribute accepts/requires a different number of arguments.
For a list of attributes and additional information on them, see the [attribute] page.
```

title: Slice grammar
---

This page describes the grammatical rules, usage, and restrictions of the Slice language.

## Slice elements

### Slice files

The `SliceFile` grammar rule is the parser's entry point.
Once a file's contents have been [preprocessed], the resulting text is matched against this rule, in its entirety.

While the contents of a Slice file may reference definitions in other files, each file is parsed separately.
There is no notion of 'including' the contents of one file in another, and errors in one file do not affect the parsing of other files.

> A slice file consists of any number of [file attributes][attribute] and up to one [mode statement][mode-statement].

Reviewer 2: ```suggestion
A slice file consists of any number of [file attributes][attribute] and at most one [mode statement][mode-statement].
```

Think this might read a bit better.

> These can appear in any order, but specifying more than one mode in a single file is forbidden.

Reviewer 2: ```but specifying more than one mode in a single file is forbidden.```

Is this necessary, we say in the above line that you can only have 0 or 1 mode statement.

Author: It's not necessary, I just wrote it this way so it would read better than "these can appear in any order".
I rewrote the sentence to:
```
File attributes can appear before or after (inclusively) the mode statement.
```

> This is followed by an optional [module declaration][module-declaration], followed by any number of Slice definitions.

Reviewer: I agree, this should be written in a clearer manner.

> This is followed by an optional [module declaration][module-declaration], followed by any number of Slice definitions.

Author: I added this sentence to the paragraph:
`It is illegal for a file to contain Slice definitions without a module declaration.`

We also state this in the "module declaration" section, but slightly different:
`All Slice definitions must be contained within modules - Slice definitions cannot exist at global scope.`

```ebnf {% showTitle=false %}
SliceFile
    : SliceFilePrelude Module? Definition*
    ;

SliceFilePrelude
    : EMPTY
    | SliceFilePrelude FileCompilationMode
    | SliceFilePrelude FileAttribute
    ;

Definition
    : Struct | Class | Exception | Interface | Enum | CustomType | TypeAlias
    ;
```

For additional information on Slice files, see the [Slice file][slice-file-guide] page.

### Mode declarations

Mode declarations consist of the `mode` keyword, followed by an equals sign, followed by an identifier.

> That identifier must be a valid Slice compilation mode - either `Slice1` or `Slice2`.

Reviewer 2: ```suggestion
This identifier must be a valid Slice compilation mode - either `Slice1` or `Slice2`.
```

```ebnf {% showTitle=false %}
FileCompilationMode
    : "mode" "=" Identifier
    ;
```

Mode declarations must come before any Slice definitions or [module declarations][module-declaration].
While the compilation mode you select will affect which Slice features are available to you,
it has no effect on the parsing of, or validity of syntax within, your Slice file.

> For example, even though classes are a Slice1-only feature, `class` is still a keyword within a Slice2 file,
> and if you define a class, syntax errors in that class's definition will still be reported normally.

Reviewer 2: ```suggestion
For example, even though classes are a Slice1-only feature, `class` is still a keyword within a Slice2 file. As such, if you define a class in a Slice2 file, syntax errors in that class's definition are still reported.
```
 
Maybe something like this .

Author: I'm open to re-wording this section, but I don't think "as such" makes sense for it.
It implies that `class being a keyword in Slice2` => `syntax errors are still reported for it`.
They're 2 separate examples of the same concept, one doesn't really follow from the other.

For additional information on modes, see the [compilation mode][compilation-mode-guide] page.

### Module declarations

Module declarations consist of the `module` keyword, followed by an [identifier].

> [local attributes][attribute] can be applied to a module in its prelude, but [doc comments][doc-comment] on modules are forbidden.

Author: Good catch, fixed.

    ;
```

If you are using an RPC framework, every interface definition also defines a _proxy type_.
This proxy type shares the same name and scope as its corresponding interface and can be used as a type.

```slice
interface Foo { // Defines an interface named `Foo`.
    op() -> Foo // Returns a proxy named `Foo`.
}
```

It is illegal to use an interface as a type without an RPC framework.

For additional information on interfaces and proxies, see the [interface][interface-guide] and [proxy][proxy-guide] pages.

### Operations

An operation consists of an [identifier], followed be a list of [parameters][parameter], optionally followed by a return type, optionally followed by an [exception specification][exception-specification-guide].
Operations support the `idempotent` modifier keyword. For information about its effects, see the [idempotent][idempotent-guide] section.
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the operation in its prelude.

```ebnf {% showTitle=false %}
Operation
    : Prelude "idempotent"? Identifier "(" UndelimitedList<Parameter> ")" ReturnType? ExceptionSpecification?
    ;
```

> There are two syntaxes for specifying a return type, depending on the number of elements you're returning.

Reviewer: I think return types should be documented as part of Parameters, not within Operation(s).

The syntax for return parameters is largely the same as the syntax for operation parameters, and the Language Guide describe them together:
https://docs.icerpc.dev/slice2/language-guide/parameters

Author: I tried to do this, but it's awkward.
Yes, they're very similar to parameters, but the grammar rules are totally separate.

The 'special' parameter syntax is baked into the rule for `ReturnType`s:
```ebnf {% showTitle=false %}
ReturnType
    : "->" Tag? "stream"? TypeRef
    | "->" "(" UndelimitedList<Parameter> ")"
    ;
```

And I think it makes the most sense to document the concept of "return types" on operations, instead of parameters.

Note that we have a sentence and link mentioning this in the documentation though:
`The syntax of a single return type is identical to a [parameter], but without an identifier or colon separator.`

Reviewer: > And I think it makes the most sense to document the concept of "return types" on operations, instead of parameters.

So the documentation for operations includes the description for return types and references the documentation for parameters (which are at the same level as operations) even though both parameters and return types are used only in operations.

This definitely makes complete sense!

Author: > even though both parameters and return types are used only in operations.

Yes, and enumerators are only used in enums, what's your point?
The documentation follows the grammar. As much as you seem to dislike it, the grammar of single return types and parameters _is different_.

> which are at the same level as operations

You really need to make up your mind about whether I should continue with this PR as-is or restructure it.

A single return type can be specified by an arrow, followed by an optional [tag], followed by an optional [`stream`][stream-guide] keyword, and then a [type].
The syntax of a single return type is identical to a [parameter], but without an identifier or colon separator.
Multiple return parameters can be specified as a tuple. This consists of an arrow, followed by a list of [parameters][parameter] wrapped in parenthesis.
It is illegal for a return tuple to contain less than 2 elements. 

```ebnf {% showTitle=false %}
ReturnType

>     : "->" Tag? "stream"? TypeRef

Author: I agreed with your reasoning, so I went to make this change, and remembered why we do it this way.
Splitting it into two rules causes an ambiguity in the grammar:
```
"->" Tag? TypeRef
"-> "stream"? TypeRef
```
If all the parser sees is (`"->" TypeRef`), both rules would be equally applicable, and it couldn't know which to call.

>     : "->" Tag? "stream"? TypeRef

Author: We could split it into 3 non-optional rules I suppose:
```
"->" Tag TypeRef
"-> "stream" TypeRef
"->" TypeRef
```

    | "->" "(" UndelimitedList<Parameter> ")"
    ;
```

> There are two syntaxes for [exception specifications][exception-specification-guide], depending on the number of [exceptions][exception] being specified.

Author: I added this line to the "Exception Specification" part:
```
Exception specifications can only declared in [`Slice1`][compilation-mode-guide] mode.
```

A single exception can be specified with the `throws` keyword, followed by the (possibly scoped) identifier of an exception.
Multiple exceptions can be specified as a tuple. This consists of the `throws` keyword, followed by the (possibly scoped) identifiers of one or more exceptions, separated by commas.

```ebnf {% showTitle=false %}
ExceptionSpecification
    : "throws" TypeRef
    | "throws" "(" NonEmptyCommaList<TypeRef> ")"
    ;
```

> Parameter lists may be optionally separated by commas. This applies to both operation parameters and return parameters.

Author: Good point, I moved it up, so it is the last sentence in the "Return Parameters" section

For additional information on operations, see the [operation][operation-guide] pages.

### Parameters

The syntax for parameters is identical to the syntax for [fields][field], but with added support for the `stream` modifier keyword.

Parameters consist of an [identifier], followed by a colon, and then a [type].
Optionally, a [tag] may be applied to the parameter, directly before its identifier, making this a [tagged parameter][tag].
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the field in its prelude.

```ebnf {% showTitle=false %}
Parameter
    : Prelude Tag? Identifier ":" "stream"? TypeRef
    ;
```

> Parameters also support the [`stream`][stream-guide] modifier keyword, which may be applied to the parameter's type.

Author: I added this to the 'parameters' section where we talk about streams:
```
Streamed parameters are only supported in [`Slice2`][compilation-mode-guide] mode, and within a list of parameters,
at most one may be streamed and it must be the last parameter in that list.
```

For additional information on parameters, see the [parameter][parameter-guide] pages.

### Enum types

Enum definitions consist of the `enum` keyword, followed by an [identifier], optionally followed by an [underlying type][type], and then the enum's body.
Underlying types are specified by a single colon, followed by a type.

> Underlying types must be integral and non-optional. Enums have a default underlying type of [varint32][primitive] if none is specified.

Author: Yep, it's true. That's the default value for the `underlying_type` field of an `Enum`.

But you're right that we require an underlying type for Slice2 enums, however this isn't really a _grammar_ thing.
It's a temporary rule that will let us add enums with associated values without breaking people's enums.

> Underlying types must be integral and non-optional. Enums have a default underlying type of [varint32][primitive] if none is specified.

Author: I removed this line because it's probably just confusing.
Once we add enums with associated values, this section will be rewritten anyways.

Enum bodies consist of a list of [enumerators][enumerator] wrapped in a pair of braces. These enumerators may be optionally separated by commas.
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the enum in its prelude.

```ebnf {% showTitle=false %}
Enum
    : Prelude "unchecked"? "enum" Identifier (":" TypeRef)? "{" UndelimitedList<Enumerator> "}"
    ;
```

Enums also support the `unchecked` modifier keyword.
An [unchecked-enum][unchecked-enum-guide] can be empty. An enum without this modifier must have at least one enumerator.

For additional information on enums, see the [enum][enum-guide] page.

### Enumerators

Enumerators can only be defined within [enums][enum].
Enumerator definitions consist of an [identifier], optionally followed by an enumerator value.

> Enumerator values consist of an equals sign, followed by a (possibly signed) [integer].

Author: I can try to reword it more clearly, but formally, it is correct.
Sentence 2 says an "enumerator value" is optional.
And sentence 3 says "an enumerator value is '=' followed by a number", no optional qualifiers.

So you can leave off the whole value part, but you can't just have half of it (only the equals sign).

Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the enumerator in its prelude.

```ebnf {% showTitle=false %}
Enumerator
    : Prelude Identifier ("=" SignedInteger)?
    ;
```

> The enum that an enumerator is defined within determines the allowed values;

Author: Changed as suggested.

```

Type aliases can be used anywhere a type can be used, and behave identically to their underlying type.
It is legal for type aliases to alias other type aliases.

For additional information on type aliases, see the [type alias][type-alias-guide] page.

### Type references

Type references consist of any number of [type attributes][attribute], followed a type, optionally followed by a question mark.

Built-in types like [primitives][primitive], [sequences][sequence], and [dictionaries][dictionary] can be referenced by their keywords.
User defined types can be referenced by their (possibly scoped) [identifiers][identifier].

```ebnf {% showTitle=false %}
TypeRef
    : LocalAttribute* TypeRefDefinition "?"?
    ;

TypeRefDefinition
    : Primitive | Sequence | Dictionary | RelativeIdentifier | GlobalIdentifier
    ;
```

### Tags

Tags consist of the `tag` keyword, followed by a positive integer wrapped in a pair of parenthesis.
This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).
Tags can only be applied to [fields][field] and [parameters][parameter], and only if the type of that field/parameter:
- is optional (it ends with `?` symbol)

> - is not a class, and does not use classes internally

Author: I know this is super pedantic, but I think pedantry matters in the reference.

This change would let through:
```
compact struct Foo {
    a: AnyClass
}
```
'Foo' `is not a class`, and `does not reference a type that contains classes` (`AnyClass` doesn't contain a class).

```ebnf {% showTitle=false %}
Tag
    : "tag" "(" SignedInteger ")"
    ;
```

> ### Identifiers

Author: I'm leaving the order and structure of the reference until after this PR.

title: Slice grammar
---

This page describes the grammatical rules, usage, and restrictions of the Slice language.

## Slice elements

### Slice files

The `SliceFile` grammar rule is the parser's entry point.
Once a file's contents have been [preprocessed], the resulting text is matched against this rule, in its entirety.

While the contents of a Slice file may reference definitions in other files, each file is parsed separately.
There is no notion of 'including' the contents of one file in another, and errors in one file do not affect the parsing of other files.

A slice file consists of any number of [file attributes][attribute] and at most one [mode statement][mode-statement].
File attributes can appear before or after (inclusively) the mode statement.
This is followed by an optional [module declaration][module-declaration], followed by any number of Slice definitions.

> It is illegal for a file to contain Slice definitions without a module declaration.

Reviewer 2: I know the intent of my original comment on this first ilegal comment got muddled so for clarity I think this should be changed too.

Once a file's contents have been [preprocessed], the resulting text is matched against this rule, in its entirety.

While the contents of a Slice file may reference definitions in other files, each file is parsed separately.
There is no notion of 'including' the contents of one file in another, and errors in one file do not affect the parsing of other files.

A slice file consists of any number of [file attributes][attribute] and at most one [mode statement][mode-statement].
File attributes can appear before or after (inclusively) the mode statement.
This is followed by an optional [module declaration][module-declaration], followed by any number of Slice definitions.
It is illegal for a file to contain Slice definitions without a module declaration.

```ebnf {% showTitle=false %}
SliceFile
    : SliceFilePrelude Module? Definition*
    ;

SliceFilePrelude
    : EMPTY
    | SliceFilePrelude FileCompilationMode
    | SliceFilePrelude FileAttribute
    ;

Definition
    : Struct | Class | Exception | Interface | Enum | CustomType | TypeAlias
    ;
```

For additional information on Slice files, see the [Slice file][slice-file-guide] page.

### Mode declarations

> Mode declarations consist of the `mode` keyword, followed by an equals sign, followed by an identifier.

Author: I added a link as suggested.

This identifier must be a valid Slice compilation mode - either `Slice1` or `Slice2`.

```ebnf {% showTitle=false %}
FileCompilationMode
    : "mode" "=" Identifier
    ;
```

Mode declarations must come before any Slice definitions or [module declarations][module-declaration].
The compilation mode you select determines which Slice features are available to you
but has no effect on the parsing of, or validity of syntax within, your Slice file.
For example, even though classes are a Slice1-only feature, `class` is still a keyword within a Slice2 file,

> and if you define a class, syntax errors in that class's definition will still be reported normally.

Author: I've reworded it to not talk about "reporting syntax errors":
```
The compilation mode you select determines which Slice features are available to you but has no effect on the parsing
of your Slice files. For example, classes are a Slice1-only feature, so it is invalid to define one in a Slice2 file.
But within this file, `class` is still treated as a keyword, and class definitions must still be syntactically correct.
```

For additional information on modes, see the [compilation mode][compilation-mode-guide] page.

### Module declarations

Module declarations consist of the `module` keyword, followed by an [identifier].

> [Local attributes][attribute] can be applied to a module in its prelude, but [doc comments][doc-comment] on modules are forbidden.

Reviewer 2: Same here `[doc comments][doc-comment]`

```ebnf {% showTitle=false %}
Module
    : Prelude "module" RelativeIdentifier
    ;
```

If the module's identifier is [unscoped][identifier], this declares a _top-level module_.
If the module's identifier is [scoped][identifier], this declares a _submodule_.

Submodules are modules that are logically contained within another module (called the _parent_).

> Slice definitions within a submodule can reference types in parent modules without qualification.

Author: I was worried it'd be confused with the "Slice definitions" that starts the sentence.
But you're right, this statement leaves out exceptions.
Changed to `can reference other definitions`.

```slice
module A // Declares a top-level module named `A`.
struct AStruct {}
```

```slice
module A::B // Declares a submodule named `B` that is contained within module `A`.
typealias T = AStruct // Don't need to qualify as `A::AStruct`.
```

All Slice definitions must be contained within modules - Slice definitions cannot exist at global scope.

> If a Slice file contains no Slice definitions, it is legal to omit the module declaration.

Author: If you want to, I'll mention it.
But I think it's already kind of obvious.

I'm pretty sure most languages allow empty files as valid input.
I guess we could check if any of them mention it explicitly.

> If a Slice file contains no Slice definitions, it is legal to omit the module declaration.

Reviewer 3: or just remove "If a Slice file contains no Slice definitions, it is legal to omit the module declaration."

> If a Slice file contains no Slice definitions, it is legal to omit the module declaration.

Author: I removed it, like you suggested. Now we just say:
```
If a slice file contains any Slice definitions, a module declaration is required, ...
```
A reader is free to infer the contrapositive, but we don't explicitly say it anymore.

Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,
and any Slice definitions in the remainder of the file will be contained within the declared module.

Module declarations must come before any Slice definitions and must come after any [file attributes][attribute] or [mode declarations][mode-declaration].
There can be at most one module declaration per Slice file.

```ebnf {% showTitle=false %}
SliceFile
    : SliceFilePrelude Module? Definition*
    ;
```

> It is legal for multiple Slice files to declare the same module.

Author: Sure.

Slice definitions defined in those files will be in the same module, even if they're in different Slice files.

For additional information on modules, see the [module][module-guide] page.

### Primitive types

> Slice supports 17 primitive types. These types are always available - they're defined in any scope of any Slice file -

Author: So replace the whole 2 sentences with "Slice supports the following primitive types:"?

> I find this sentence confusing. What are you trying to explain here?

That the types are built-in, and don't require any kind of scoping to use.
Maybe this too obvious and everyone will already expect this of 'primitive' types.

Slice supports 2 built-in collection types: sequences and dictionaries.

These types are generic over the kinds of elements they contain.
Sequences require a single type parameter: the _element_ type.
Dictionaries require two type parameters: the _key_ and _value_ types.

Any Slice type can be specified for element and value types, but since dictionary keys must be comparable, only the following are valid key types:
- `bool`
- `string`
- integral types (`int8`, `uint8`, `int16`, `uint16`, `int32`, `uint32`, `varint32`, `varuint32`, `int64`, `uint64`, `varint62`, `varuint62`)
- enum types
- custom types
- compact structs whose fields are also valid key types

Sequences and dictionaries can be referenced with their respective keywords (`Sequence` and `Dictionary`),
but must have concrete types specified for their type parameters.

```ebnf {% showTitle=false %}
Sequence
    : "Sequence" "<" TypeRef ">"
    ;

Dictionary
    : "Dictionary" "<" TypeRef "," TypeRef ">"
    ;
```

```slice
Sequence<bool>
Dictionary<uint8, string>
```

For additional information on sequences and dictionaries, see the [sequence][sequence-guide] and [dictionary][dictionary-guide] pages.

### Struct types

> Struct definitions consist of the `struct` keyword, followed by an [identifier], and then the struct's body.

Author: Unlike our other pages, I don't think users will read this one linearly (from the top to the bottom).
They'll probably just jump to the section that's relevant to them, skipping anything before it.

So instead of a link appearing the first time on the _page_, I added a link the first time it appears in a _section_.

If they want to read about 'identifiers', making users guess where they first appear, and scroll up to it hurts navigation.

Struct bodies consist of a list of [fields][field] wrapped in a pair of braces. These fields may be optionally separated by commas.

> Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the struct in its prelude.

Reviewer 2: Same here for `[doc-comment]`

```ebnf {% showTitle=false %}
Struct
    : Prelude "compact"? "struct" Identifier "{" UndelimitedList<Field> "}"
    ;
```

Structs also support the `compact` modifier keyword.
A [compact struct][compact-struct-guide] cannot contain [tagged fields][tag].

For additional information on structs, see the [struct][struct-guide] page.

### Class types

Classes can only be defined or referenced in [`Slice1`][compilation-mode-guide] mode.

> Classes can never be tagged when used as the type of a [field] or [parameter].

Author: I added a link as suggested.

Class definitions consist of the `class` keyword, followed by an [identifier], optionally followed by a [compact ID][compact-id-guide], optionally followed by a base class, and then the class's body.
Compact IDs consist of a positive integer wrapped in a pair of parenthesis.

> This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).

Reviewer 2: ```suggestion
This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32 bit integer).
```
Should this have a space?

Author: Looking online, the convention seems to be using a hyphen. Fixed.

> Base classes are specified by a single colon, followed by a class's (possibly scoped) identifier.

Reviewer 2: I find that we use parthenses to mention the scoping quiet often. Could we just ommit these in multiple places? Would make it a bit nicer to read.

Author: Omit the parenthesis or the entire thing? I'm open to alternate wordings (including those lacking parenthesis), but I do think we should be clear that both scoped and unscoped identifiers are valid.

    : Prelude "exception" Identifier (":" TypeRef)? "{" UndelimitedList<Field> "}"
    ;
```

For additional information on exceptions, see the [exception][exception-guide] page.

### Fields

Fields can only be declared within [structs][struct], [classes][class], and [exceptions][exception].
Field declarations consist of an [identifier], followed by a colon, and then a [type].
Optionally, a [tag] may be applied to the field, directly before its identifier, making this a [tagged field][tag].
Additionally, [local attributes][attribute] and a [doc-comment][doc-comment] may be applied to the field in its prelude.

```ebnf {% showTitle=false %}
Field
    : Prelude Tag? Identifier ":" TypeRef
    ;
```

For additional information on fields, see the [field][field-guide] page.

### Interfaces

Interface definitions consist of the `interface` keyword, followed by an [identifier], optionally followed by a list of base interfaces, and then the interface's body.
Base interfaces are specified by a single colon, followed by the (possibly scoped) identifiers of one or more interfaces, separated by commas.
Interface bodies consist of a set of [operations][operation] wrapped in a pair of braces.

> Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the class in its prelude.

Author: Good catch!

```ebnf {% showTitle=false %}
Interface
    : Prelude "interface" Identifier (":" NonEmptyCommaList<TypeRef>)? "{" Operation* "}"
    ;
```

> If you are using an RPC framework, every interface definition also defines a _proxy type_.

Author: I'm fine removing this whole paragraph on proxies. I also find it out of place, and frankly confusing.

I only added it because we have an entire page for proxies in the language guide, and we draw a big line between interfaces and proxies. Should we do anything about that page?

This proxy type shares the same name and scope as its corresponding interface and can be used as a type.

```slice
interface Foo { // Defines an interface named `Foo`.
    op() -> Foo // Returns a proxy named `Foo`.
}
```

> It is illegal to use an interface as a type without an RPC framework.

Reviewer 2: I think we should avoid the term `illegal`. Whenever I read it I always think "well its not really illegal, im not going to go to prison for it etc.". I think something more like "disallowed", "prohibited", etc. generally fits better. Illegal generally invokes the idea of the actual judicial system and law to me.

We use illegal in three spots total and this comment applies to all three.

For example:

```suggestion
Using an interface as a type is prohibited without an RPC framework.
```

Author: I've had the same thought before, but over the years it's left me.
'legal' and 'illegal' are pretty common in formal specifications (including things like criminal statutes),
so I don't think it's a poor fit for the reference. But I'm also not married to it.

How about:
`Interfaces cannot be used as types without an RPC framework.`? Simple and short.
Even if it's _technically_ untrue, because you _can_ do anything. It just might not compile...

Generally, I find `X is <past-tense-verb-with-ed>` out of place in specifications.
The specification isn't taking any actions, it's a passive list of statements.
I'm sure I haven't consistently stuck to that myself though :)

Reviewer: The core issue here is not illegal vs cannot. This statement is incorrect: the Slice language and the Slice parser don't care and shouldn't care about RPC frameworks. 

The RPC framework aspect is purely about code generation. And that's outside the scope of the Slice language/parser.

Author: I removed this whole paragraph talking about proxies. I only added it because the language guide tries to draw a big line between interfaces (non-types) and proxies (types). But I always felt it was out of place here.

Now we just have interfaces and we call them types. At the language level, this is really what it looks like anyways.
We should communicate somewhere that this is not allowed unless you have an RPC framework though...

> For additional information on interfaces and proxies, see the [interface][interface-guide] and [proxy][proxy-guide] pages.

Author: I just removed the whole paragraph about proxies. Now we just say interfaces are a type.
That's what it looks like from the language's perspective anyways.

### Operations

An operation consists of an [identifier], followed be a list of [parameters][parameter], optionally followed by a return type, optionally followed by an [exception specification][exception-specification-guide].
Operations support the `idempotent` modifier keyword. For information about its effects, see the [idempotent][idempotent-guide] section.
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the operation in its prelude.

```ebnf {% showTitle=false %}
Operation
    : Prelude "idempotent"? Identifier "(" UndelimitedList<Parameter> ")" ReturnType? ExceptionSpecification?
    ;
```

There are two syntaxes for specifying a return type, depending on the number of elements you're returning.

> A single return type can be specified by an arrow, followed by an optional [tag], followed by an optional [`stream`][stream-guide] keyword, and then a [type].

Reviewer 2: [stream-guide] link not working.

There are two syntaxes for [exception specifications][exception-specification-guide], depending on the number of [exceptions][exception] being specified.
A single exception can be specified with the `throws` keyword, followed by the (possibly scoped) identifier of an exception.
Multiple exceptions can be specified as a tuple. This consists of the `throws` keyword, followed by the (possibly scoped) identifiers of one or more exceptions, separated by commas.
Exception specifications can only declared in [`Slice1`][compilation-mode-guide] mode.

```ebnf {% showTitle=false %}
ExceptionSpecification
    : "throws" TypeRef
    | "throws" "(" NonEmptyCommaList<TypeRef> ")"
    ;
```

For additional information on operations, see the [operation][operation-guide] pages.

### Parameters

The syntax for parameters is identical to the syntax for [fields][field], but with added support for the `stream` modifier keyword.

Parameters consist of an [identifier], followed by a colon, and then a [type].
Optionally, a [tag] may be applied to the parameter, directly before its identifier, making this a [tagged parameter][tag].

> Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the field in its prelude.

Reviewer 3: ```suggestion
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the parameter in its prelude.
```

```ebnf {% showTitle=false %}
Parameter
    : Prelude Tag? Identifier ":" "stream"? TypeRef
    ;
```

> Parameters also support the [`stream`][streamed-parameters-guide] modifier keyword, which may be applied to the parameter's type.

Reviewer 3: > Parameters also support the [stream][streamed-parameters-guide] modifier keyword,

It seems redundant to mention the support for the 'stream' modifier keyword here since it's already stated in the previous paragraph.

Author: I simplified the first sentence to just say `..., but with added support for streams`
Instead of re-using the exact same terminology we use down here when we explain it in greater detail.

For additional information on enums, see the [enum][enum-guide] page.

### Enumerators

Enumerators can only be defined within [enums][enum].
Enumerator definitions consist of an [identifier], optionally followed by an enumerator value.
Enumerator values consist of an equals sign, followed by a (possibly signed) [integer].
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the enumerator in its prelude.

```ebnf {% showTitle=false %}
Enumerator
    : Prelude Identifier ("=" SignedInteger)?
    ;
```

The enum that an enumerator is defined within determines the allowed values.
An enumerator's value must be within the bounds of the enum's underling type.

For additional information on enumerators, see the [enumerator][enum-guide] page.

### Custom types

Custom type definitions consist of the `custom` keyword, followed by an [identifier].
Additionally, [local attributes][attribute] and a [doc-comment] may be applied to the custom type in its prelude.

```ebnf {% showTitle=false %}
CustomType
    : Prelude "custom" Identifier
    ;
```

> For a custom type to be mappable, it must have a `[x::type]` attribute applied to it, where `x` is the mapping language's prefix.

Reviewer 3: Consider rephrasing the expression 'to be mappable' for clarity. How about:

For a custom type to be usable with a particular mapping language, it should have an attribute in the format [x::type], where x represents the prefix of that language.

Author: I rewrote it like you said, with some minor changes:
```
For a custom type to be usable with a particular language mapping, it must have an attribute of the form `[x::type]`,
where `x` is the Slice prefix of that language.
```

Let me know if you think it needs any more changes.

This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).
Tags can only be applied to [fields][field] and [parameters][parameter], and only if the type of that field/parameter:
- is optional (it ends with `?` symbol)
- is not a class, and does not use classes internally

```ebnf {% showTitle=false %}
Tag
    : "tag" "(" SignedInteger ")"
    ;
```

### Identifiers

Slice supports three forms of identifiers:
- An unscoped identifier. These consist of a letter, followed by any number of letters, digits, or underscores.
- A relatively scoped identifier. These consist of one or more unscoped identifiers, separated by `::` tokens.
- A globally scoped identifier. These consist of a `::` token, followed by a relatively scoped identifier.

```ebnf {% showTitle=false %}
Identifier
    : identifier
    ;

RelativeIdentifier
    : identifier ("::" identifier)*
    ;

GlobalIdentifier
    : ("::" identifier)+
    ;
```

Examples of valid identifiers:

```slice
// Examples of unscoped identifiers
foo9
b_a_r
BAZ

// Examples of relatively scoped identifiers
foo9
foo9::b_a_r::BAZ

// Examples of globally scoped identifiers
::foo9
::foo9::b_a_r::BAZ
```

> All Slice [keywords] satisfy the definition of an unscoped identifier.

Reviewer 2: `[keywords]` Link not working

These underscores carry no semantic meaning and are discarded by the parser.

Examples of valid integer literals:

```slice
123
0xFF
0b1010

335_445_996
0x_ab_cd_ef  // Same as 0xabcded
0b0_______1  // Same as 0b01

0x_0b1101  // hexadecimal literal of value 725249

// All have a value of 0
0
0_00_0
0b000
0x0
```

{% callout type="note" %}
Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
{% /callout %}

The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.

```ebnf {% showTitle=false %}
Integer
    : integer_literal
    ;

SignedInteger
    : Integer
    | "-" Integer
    ;
```

### String literals

String literals consist of any number of UTF8 characters, wrapped in a pair of double quotes.
Any characters within a string literal are treated as opaque text by the parser.

Slice supports escaping special characters within string literals.
This allows you to include unprintable characters or literal double quote characters in your string.

> Characters can be escaped by prefixing them with a backslash.

Reviewer 3: > Characters can be escaped by prefixing them with a backslash.

Seems confusing `\t` is this escaping the character 't' or the tab character?

Author: I've now clarified this. `\t` is escaping the character `t`.
I also added the following example, just to remove any shred of ambiguity:
```rust
"the letter n: '\n'"  // the letter n: 'n'
```

To write a literal backslash, you must escape it with another backslash: `\\`.

The parser resolves any escaped characters, removing the `\` escape character where present.
Apart from this, no additional processing is performed on string literals.

Examples of valid string literals:

```slice
""
"hello world"
"the variable \"foo\" doesn't exist"
"a backslash: '\\'"
```

String literals can only appear as arguments to an [attribute].

### Attributes

An attribute consists of a directive, optionally followed by attribute arguments.

A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of specifying namespaces, the presence of scopes limits when the attribute is applicable.
When no scopes are specified, the attribute is always applied, but when scopes are present, it is applied conditionally.

> For example, `deprecated` is always applied, but `cs::type` will only be applied when compiling to C#.

Reviewer 3: This is somehow up to compiler implementation, from the language point of view is just a scope for the attribute.

Author: I just removed the last 2 sentences completely, since we couldn't find a wording that ticked all the boxes.
Now we just say:
```
A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of specifying namespaces,
the presence of scopes limits when the attribute is applied.
```

    : (T comma?)*
    ;
```

[slice-file]: #slice-files
[module-declaration]: #module-declarations
[mode-declaration]: #mode-declarations
[primitive]: #primitive-types
[sequence]: #collection-types
[dictionary]: #collection-types
[struct]: #struct-types
[class]: #class-types
[exception]: #exception
[field]: #fields
[operation]: #operations
[parameter]: #parameters
[enum]: #enum-types
[enumerator]: #enumerators
[tag]: #tag
[identifier]: #identifiers
[attribute]: #attributes
[type]: #type-references
[integer]: #integer-literals
[string]: #string-literals

> [slice-file-guide]: /slice/basics/slice-files

Author: I'm fine removing them, but why not?
In my mind they'd both link to each other.

The guide is for explaining the motivation and idea of something,
and the reference is for explaining what _exactly_ is this something.
They're documenting two different sides of the same thing.
